test/mockInterfaces.go

```diff
@@ -0,0 +1,302 @@
+}
+
+func (h *TestHostService) GetDNSRecord(ctx context.Context, subDomain string, managedZone *v1alpha1.ManagedZone, _ v1.Object) (*v1alpha1.DNSRecord, error) {
+	if subDomain == FailFetchDNSsubdinain {
+		return &v1alpha1.DNSRecord{}, fmt.Errorf(FailFetchDNSsubdinain)
+	}
+	record := &v1alpha1.DNSRecord{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      managedZone.Spec.DomainName,
+			Namespace: managedZone.GetNamespace(),
+		},
+	}
+
+	if err := h.controlClient.Get(ctx, client.ObjectKeyFromObject(record), record); err != nil {
+		return nil, err
+	}
+	return record, nil
+}
+
+func (h *TestHostService) GetManagedZoneForHost(ctx context.Context, host string, t traffic.Interface) (*v1alpha1.ManagedZone, string, error) {
+	hostParts := strings.SplitN(host, ".", 2)
+	if len(hostParts) < 2 {
+		return nil, "", fmt.Errorf("unable to parse host : %s on traffic resource : %s", host, t.GetName())
+	}
+	subDomain := hostParts[0]
+	zone := v1alpha1.ManagedZoneList{}
+
+	err := h.controlClient.List(ctx, &zone, client.InNamespace(t.GetNamespace()))
+	if err != nil {
+		return nil, "", err
+	}
+	if len(zone.Items) == 0 {
+		return nil, subDomain, nil
+	}
+	return &zone.Items[0], subDomain, nil
+}
+
+func (h *TestHostService) AddEndpoints(_ context.Context, gateway traffic.Interface, _ *v1alpha1.DNSRecord) error {
+	hosts := gateway.GetHosts()
+	for _, host := range hosts {
+		if host == FailEndpointsHostname {
+			return fmt.Errorf(FailEndpointsHostname)
+		}
+	}
+	return nil
+}
+
+func NewTestHostService(client client.Client) *TestHostService {
+	return &TestHostService{controlClient: client}
+}
+
+type TestGatewayPlacer struct{}
+
+func (p *TestGatewayPlacer) Place(_ context.Context, upstream *v1beta1.Gateway, _ *v1beta1.Gateway, _ ...v1.Object) (sets.Set[string], error) {
+	if upstream.Labels == nil {
+		return nil, nil
+	}
+	if *upstream.Spec.Listeners[0].Hostname == FailPlacementHostname {
+		return nil, fmt.Errorf(FailPlacementHostname)
+	}
+	targetClusters := sets.Set[string](sets.NewString())
+	targetClusters.Insert(TestCluster)
+	return targetClusters, nil
+}
+
+func (p *TestGatewayPlacer) GetPlacedClusters(_ context.Context, gateway *v1beta1.Gateway) (sets.Set[string], error) {
+	if gateway.Labels == nil {
+		return nil, nil
+	}
+	placedClusters := sets.Set[string](sets.NewString())
+	placedClusters.Insert(TestCluster)
+	return placedClusters, nil
+}
+
+func (p *TestGatewayPlacer) GetClusters(_ context.Context, _ *v1beta1.Gateway) (sets.Set[string], error) {
+	return nil, nil
+}
+
+func (p *TestGatewayPlacer) ListenerTotalAttachedRoutes(_ context.Context, _ *v1beta1.Gateway, listenerName string, _ string) (int, error) {
+	if listenerName == ValidTestHostname {
+		return 1, nil
+	}
+	return 0, nil
+}
+
+func (p *TestGatewayPlacer) GetAddresses(_ context.Context, _ *v1beta1.Gateway, _ string) ([]v1beta1.GatewayAddress, error) {
+	t := v1beta1.IPAddressType
+	return []v1beta1.GatewayAddress{
+		{
+			Type:  &t,
+			Value: "1.1.1.1",
+		},
+	}, nil
+}
```

Comment:
This ones better, but i think it needs to be more explicit where it's being used what values are going to actually be returned if you are trying to make this into a reusable interface that any test could use where they want to mimic the placement functionality but not actually need to use it.

I had to create a similar one here https://github.com/Kuadrant/multicluster-gateway-controller/pull/237/files#diff-525abdb709ea5f8ab96976a27ec21c0e9545dd47dd82b0e2abff7a475f9672e9R40